can/interfaces/multicast_ip/bus.py

```diff
@@ -0,0 +1,327 @@
+import logging
+import select
+import socket
+import struct
+
+from typing import List, Optional, Tuple, Union
+
+log = logging.getLogger(__name__)
+
+import can
+from can import BusABC
+from can.typechecking import AutoDetectedConfig
+from can.typechecking import ReadableBytesLike
+
+from .utils import pack_message, unpack_message, check_msgpack_installed
+
+
+# see socket.getaddrinfo()
+IPv4_ADDRESS_INFO = Tuple[str, int]  # address, port
+IPv6_ADDRESS_INFO = Tuple[str, int, int, int]  # address, port, flowinfo, scope_id
+IP_ADDRESS_INFO = Union[IPv4_ADDRESS_INFO, IPv6_ADDRESS_INFO]
+
+# Additional constants for the interaction with Unix kernels
+SO_TIMESTAMPNS = 35
+
+
+class MulticastIpBus(BusABC):
+    """A virtual interface that allows to communicate between multiple processes using UDP over Multicast IP.
+
+    It supports IPv4 and IPv6, specified via the channel (which really is just a multicast IP address as a
+    string). You can also specify the port and the IPv6 *hop limit*/the IPv4 *time to live* (TTL).
+
+    This bus does not support filtering based on message IDs on the kernel level but instead provides it in
+    user space (in Python) as a fallback.
+
+    Both default addresses should allow for multi-host CAN networks in a normal local area network (LAN) where
+    multicast is enabled.
+
+    .. note::
+        The auto-detection of available interfaces (see) is implemented using heuristic that checks if the
+        required socket operations are available. It then returns two configurations, one based on
+        the :attr:`~MulticastIpBus.DEFAULT_GROUP_IPv6` address and another one based on
+        the :attr:`~MulticastIpBus.DEFAULT_GROUP_IPv4` address.
+
+    .. warning::
```

Comment:
It might be good to document here the known limitations:
- no CAN id prioritization
- no guarantees on delivery / order of delivery
- no rate limiting on messages